src/condor_utils/wire_ad_buffer.cpp

```diff
@@ -0,0 +1,520 @@
+	//bool cache_lazy = (options & GET_CLASSAD_LAZY_PARSE) != 0;
+	bool fast_tricks = (options & GET_CLASSAD_FAST) != 0;
+	const size_t always_cache_string_size = 128; // no fast parse for strings > this size.
+
+	if ( ! (options & GET_CLASSAD_NO_CLEAR)) {
+		ad.Clear();
+	}
+
+	// the ad may start with the 0xCA tag that indicates a complete
+	// top level add.
+	// ad.Update wants the hashtable, so we need to scan down to it.
+	std::string_view label;
+	auto stmHash = WireClassadBuffer::find_hash(stmAd, label);
+	if (stmHash.empty()) {
+		return false;
+	}
+
+	if (time_and_flags & BINARY_CLASSAD_FLAG) {
+
+		// set a ServerTime attribute from the server time stored in the binary header
+		// This will be overwritten by ad.Update below if the binary blob has a ServerTime attribute.
+		// TODO: add an options flag for this??
+		time_t server_time = (time_t)(time_and_flags & ~BINARY_CLASSAD_MASK);
+		if (server_time && ( ! whitelist || whitelist->count(ATTR_SERVER_TIME))) {  
+			ad.Assign(ATTR_SERVER_TIME, server_time);
+		}
+
+		// peek at first pair to see if fast_tricks is possible
+		classad::ExprStream stmTmp = stmHash;
+		int numExprs = WireClassadBuffer::read_hash(stmTmp);
+		if (numExprs <= 0) return false;
+
+		std::string_view lhs;
+		binary_span expr;
+		classad::ExprTree::NodeKind kind;
+		if (WireClassadBuffer::next_pair(stmTmp, lhs, expr, kind) && kind != classad::ExprTree::EXPR_ENVELOPE) {
+			// fast tricks only work with envelope expr trees
+			fast_tricks = false;
+		}
+	}
+
+	// if we have not been asked to use fast parsing tricks
+	// we can just call the classad update from pickle functions
+	if ( ! fast_tricks) {
+		if (whitelist) {
+			return ad.Update(stmHash, *whitelist, false);
+		}
+		return ad.UpdateViaCache(stmHash);
+	}
+
+	// can't use ad.update because of fast tricks
+	// so iterate the hash here
+
+	int numExprs = WireClassadBuffer::read_hash(stmHash);
+	if (numExprs <= 0) {
+		return false;
+	}
+
+	classad::ClassAdParser parser;
+	parser.SetOldClassAd(true);
+
+	std::string attr; // TODO: remove this when we can insert into a classad froma string_view
+
+	std::string_view lhs;
+	binary_span expr;
+	classad::ExprTree::NodeKind kind;
+
+	for (int ix = 0; ix < numExprs; ++ix) {
+		if ( ! WireClassadBuffer::next_pair(stmHash, lhs, expr, kind) || lhs.empty() || expr.empty()) {
+			dprintf(D_ALWAYS, "updateAdFromRaw next_pair did not return anything to insert\n");
+			return false;
+		}
+
+		// TODO: remove this when we can insert into classads using a string_view
+		attr.assign(lhs);
+
+		// honor whitelist if any
+		if (whitelist && ! whitelist->count(attr)) {
+			continue;
+		}
+
+		classad::ExprStream stmExpr(expr);
+		if (kind != classad::ExprTree::EXPR_ENVELOPE) {
+			classad::ExprTree * tree = nullptr;
+			if (stmExpr.readNullableExpr(tree)) {
+				ad.Insert(attr, tree);
+			} else {
+				dprintf(D_ALWAYS, "updateAdFromRaw next_pair rhs is not a pickled expr\n");
```

Comment:
This is one of the few places where you log something to explain the error condition.  (That I've seen so far, anyway.)  Is this a left-over debug statement?

Reply:
yes,